### PR TITLE
Add guided GitHub secrets migration

### PR DESCRIPTION
**Pull request by lox**
## Why

GitHub's API exposes Actions secret names but not their values. Migrating them safely requires a reviewed default-branch workflow that names each source explicitly and transfers values without exposing them locally.

Depends on buildkite/buildkite#32610 for the one-use OIDC migration grant and atomic create-only batch endpoint.

## What

Add `migrate-secrets prepare` to discover repository secret names through `gh`, select a Buildkite destination through `bk`, and generate a deterministic `workflow_dispatch` workflow with a static allowlist and non-empty access policy.

Add `migrate-secrets run` to verify the exact generated workflow on the GitHub default branch, bind a 15-minute grant to its immutable repository identity and commit, and dispatch it with a non-secret grant ID. The workflow sends one in-memory HTTPS batch using GitHub OIDC; it never prints values, overwrites existing keys, or mutates repository files automatically.

This first version supports github.com repository secrets only. It excludes `GITHUB_TOKEN`, which retains the existing scoped workflow-token path.

### CODE REVIEW

File: internal/cli/secrets_migrate.go [hunk 1]
```diff
@@ -0,0 +1,748 @@
...
+          if not re.fullmatch(r"[A-Za-z0-9_-]{16,200}", grant_id):
+              sys.exit("The migration grant ID is invalid")
+
+          values = {}
+          for key, variable in secrets:
+              value = os.environ.get(variable, "")
+              size = len(value.encode("utf-8"))
+              if size == 0:
+                  sys.exit(f"GitHub secret {key} is missing or empty; no Buildkite secrets were created")
+              if size >= 8192:
```
[the remark this answers is not in the document]
Comment (lox) [reply internal/cli/secrets_migrate.go]: Fixed in 001a7b5: the generated workflow now accepts values up to the backend’s exact 32 KiB exclusive limit, with an 8 KiB regression case.

File: internal/cli/secrets_policy.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+	}
+	for ruleIndex, rule := range rules {
+		if len(rule) == 0 {
+			return "", fmt.Errorf("policy rule %d must contain at least one claim", ruleIndex+1)
+		}
+		for claim, conditions := range rule {
+			if !claims[claim] {
+				return "", fmt.Errorf("policy rule %d contains unknown claim %q", ruleIndex+1, claim)
+			}
+			if !validPolicyConditions(conditions) {
```
[the remark this answers is not in the document]
Comment (lox) [reply internal/cli/secrets_policy.go]: Fixed in 001a7b5: policy validation now applies UUID types to cluster_id, cluster_queue_id, pipeline_id, and build_creator_team, and covers the complete backend claim set.

**Comment by lox** [comment]
Addressed Buildsworth review 4943891084 in 001a7b5: aligned the generated value limit with the backend’s 32 KiB exclusive bound and validated UUID-typed policy claims during prepare. Both inline threads are resolved.

File: internal/cli/secrets_migrate.go [hunk 3]
```diff
@@ -0,0 +1,748 @@
...
+	for _, prefix := range []string{"https://github.com/", "http://github.com/", "ssh://git@github.com/", "git@github.com:"} {
+		if strings.HasPrefix(normalized, prefix) {
+			return strings.EqualFold(strings.TrimPrefix(normalized, prefix), repository)
+		}
+	}
+	return false
+}
+
+func rejectExistingBuildkiteSecrets(ctx context.Context, runner transport.Runner, cluster string, selected []string) error {
+	output, err := runner.Run(ctx, "", "bk", []string{"secret", "list", "--cluster-uuid", cluster, "--json", "--no-input"}, nil)
```
[the remark this answers is not in the document]
Comment (lox) [reply internal/cli/secrets_migrate.go]: Fixed in f6aae92: prepare now uses the organization-qualified cluster secrets API directly, paginates explicitly, and treats an empty `[]` response as a valid empty destination.

**Comment by lox** [comment]
Addressed Buildsworth review 4943904392 in f6aae92: destination preflight now uses an organization-qualified, explicitly paginated API path and supports empty clusters.

File: docs/cli.md [hunk 4]
```diff
@@ -47,6 +47,65 @@ Profile validation applies the upload trigger policy before compilation. A `not-
...
+bk auth login
+
+buildkite-gha migrate-secrets prepare \
+  --match 'DEPLOY_*' \
+  --secret API_KEY \
+  --output .github/workflows/migrate-buildkite-secrets.yml
+```
+
+The Buildkite API token used by `bk` needs `write_secrets`, and its user needs
+`manage_cluster` permission for the destination cluster.
```
[the remark this answers is not in the document]
Comment (lox) [reply docs/cli.md]: Fixed in 042ad0d: CLI docs now list the required read_pipelines, read_secrets_details, and write_secrets token scopes plus manage_cluster.

File: internal/cli/secrets_migrate.go [hunk 5]
```diff
@@ -0,0 +1,758 @@
...
+	}
+	return nil
+}
+
+func validateMigrationWorkflowPath(workflowPath string) error {
+	clean := filepath.ToSlash(filepath.Clean(workflowPath))
+	if clean != workflowPath || !strings.HasPrefix(clean, ".github/workflows/") || strings.Count(strings.TrimPrefix(clean, ".github/workflows/"), "/") != 0 {
+		return errors.New("workflow must be a clean path directly under .github/workflows")
+	}
+	extension := strings.ToLower(filepath.Ext(clean))
```
[the remark this answers is not in the document]
Comment (lox) [reply internal/cli/secrets_migrate.go]: Fixed in 042ad0d: workflow path validation now accepts only lowercase .yml and .yaml extensions, matching the grant endpoint.

File: internal/cli/secrets_migrate.go [hunk 6]
```diff
@@ -0,0 +1,758 @@
...
+jobs:
+  migrate:
+    runs-on: ubuntu-latest
+    timeout-minutes: 10
+    env:
+      DEFAULT_BRANCH: ${{ github.event.repository.default_branch }}
+      GRANT_ID: ${{ inputs.grant_id }}
+`, manifest.DefaultBranch)
+	for index, name := range manifest.SecretNames {
+		fmt.Fprintf(&output, "      MIGRATION_SECRET_%03d: ${{ secrets.%s }}\n", index, name)
```
[the remark this answers is not in the document]
Comment (lox) [reply internal/cli/secrets_migrate.go:554]: Fixed in 042ad0d: prepare caps each generated workflow at 40 secrets and directs larger migrations to multiple workflows, leaving conservative headroom under the hosted Ubuntu process environment limit.

**Comment by lox** [comment]
Addressed Buildsworth review on head f6aae92 in 042ad0d: documented all required bk scopes, matched the backend’s lowercase workflow extension contract, and capped each generated workflow at 40 secrets to stay safely below the hosted runner process environment limit. All three inline threads are resolved.

File: internal/cli/secrets_policy.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+func validateSecretsPolicy(contents string) (string, error) {
+	policy := strings.TrimSpace(contents)
+	if policy == "" {
+		return "", errors.New("policy must not be empty")
+	}
+	if strings.Contains(policy, "${{") {
+		return "", errors.New("policy must not contain GitHub expression syntax")
+	}
+	var rules []map[string]any
+	if err := yaml.Unmarshal([]byte(policy), &rules); err != nil {
```
[the remark this answers is not in the document]
Comment (lox) [reply internal/cli/secrets_policy.go]: Fixed in 2a0a81c (current head cac882a): policy validation now inspects the YAML node tree and rejects aliases before decoding, matching the backend parser.

**Comment by lox** [comment]
Addressed the Buildsworth policy-parser finding in 2a0a81c (current head cac882a) by rejecting YAML aliases before policy decoding. The thread is resolved, latest main is merged, and the full local check gate passes.

File: internal/cli/secrets_policy.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (lox) [reply internal/cli/secrets_policy.go]: Fixed in 1c50da4: prepare now rejects unquoted plain yes/no/on/off scalars and tells the user to quote them, matching Psych’s YAML 1.1 interpretation.

**Comment by lox** [comment]
Addressed the final Buildsworth parser note in 1c50da4 by rejecting ambiguous unquoted YAML 1.1 boolean spellings. Focused CLI tests and lint pass; the prior full local gate passed on the merged-main head.

File: internal/cli/secrets_migrate.go [hunk 8]
```diff
@@ -0,0 +1,762 @@
...
+	commitEndpoint := fmt.Sprintf("repos/%s/commits/%s", repository.FullName, url.PathEscape(repository.DefaultBranch))
+	commitOutput, err := runner.Run(ctx, "", "gh", []string{"api", "--jq", ".sha", commitEndpoint}, nil)
+	if err != nil {
+		return nil, "", fmt.Errorf("resolve GitHub default-branch commit with gh: %w", err)
+	}
+	commitSHA := strings.TrimSpace(string(commitOutput))
+	if len(commitSHA) != 40 {
+		return nil, "", errors.New("GitHub returned an invalid default-branch commit")
+	}
+	endpoint := fmt.Sprintf("repos/%s/contents/%s?ref=%s", repository.FullName, workflowPath, url.QueryEscape(commitSHA))
```
[the remark this answers is not in the document]
Comment (lox) [reply internal/cli/secrets_migrate.go]: Fixed in e945b38: the GitHub contents request now path-escapes the validated workflow filename segment while retaining the exact raw path in the migration grant.

**Comment by lox** [comment]
Addressed the Buildsworth workflow-path edge case in e945b38 by escaping the filename segment for the GitHub contents API while preserving the exact grant path. The inline thread is resolved; focused CLI tests and lint pass.

File: internal/cli/secrets_migrate.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (lox) [reply internal/cli/secrets_migrate.go]: Fixed in b6c3051: both API calls now use organization-relative `/clusters/...` endpoints and scope `BUILDKITE_ORGANIZATION_SLUG` to the individual `bk api` subprocess without mutating bk configuration.

File: internal/cli/secrets_policy.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (lox) [reply internal/cli/secrets_policy.go]: Fixed in b6c3051: the policy node walk now rejects custom YAML tags before decoding.

**Comment by lox** [comment]
Addressed Buildsworth review on e945b38 in b6c3051: corrected both bk API endpoint paths with subprocess-scoped organization selection, and rejected custom YAML tags. Both inline threads are resolved; focused CLI tests and lint pass.

File: internal/cli/secrets_migrate.go [hunk 9]
```diff
@@ -0,0 +1,770 @@
...
+              sys.exit("Run this migration from the repository default branch")
+          grant_id = os.environ.get("GRANT_ID", "")
+          if not re.fullmatch(r"[A-Za-z0-9_-]{16,200}", grant_id):
+              sys.exit("The migration grant ID is invalid")
+
+          values = {}
+          for key, variable in secrets:
+              value = os.environ.get(variable, "")
+              size = len(value.encode("utf-8"))
+              if size == 0:
```
[the remark this answers is not in the document]
Comment (lox) [reply internal/cli/secrets_migrate.go]: Fixed in the new head: the generated workflow now treats whitespace-only values as missing before requesting OIDC or sending the batch, with an execution test.

**Comment by lox** [comment]
Addressed Buildsworth review on b6c3051 in d9520ad: whitespace-only GitHub values now fail before OIDC or the migration request. The inline thread is resolved; focused CLI tests and lint pass.
